src/lib/proof-system/sideloaded.unit-test.ts

```diff
@@ -0,0 +1,225 @@
+      },
+    },
+  },
+});
+
+class SampleSideloadedProof extends DynamicProof<Field, Void> {
+  static publicInputType = Field;
+  static publicOutputType = Void;
+  static maxProofsVerified = 0 as const;
+}
+
+class SampleSideloadedProof2 extends DynamicProof<Program2Struct, Field> {
+  static publicInputType = Program2Struct;
+  static publicOutputType = Field;
+  static maxProofsVerified = 0 as const;
+}
+
+const sideloadedProgram = ZkProgram({
+  name: 'sideloadedProgram',
+  publicInput: Field,
+  methods: {
+    recurseOneSideloaded: {
+      privateInputs: [SampleSideloadedProof, VerificationKey],
+      async method(
+        publicInput: Field,
+        proof: SampleSideloadedProof,
+        vk: VerificationKey
+      ) {
+        proof.verify(vk);
+
+        proof.publicInput.assertEquals(publicInput, 'PublicInput not matching');
+      },
+    },
+    recurseTwoSideloaded: {
+      privateInputs: [
+        SampleSideloadedProof,
+        VerificationKey,
+        SampleSideloadedProof2,
+        VerificationKey,
+      ],
+      async method(
+        publicInput: Field,
+        proof1: SampleSideloadedProof,
+        vk1: VerificationKey,
+        proof2: SampleSideloadedProof2,
+        vk2: VerificationKey
+      ) {
+        proof1.verify(vk1);
+        proof2.verify(vk2);
+
+        proof1.publicInput
+          .add(proof2.publicInput.field1.add(proof2.publicInput.field2))
+          .assertEquals(publicInput, 'PublicInput not matching');
+      },
+    },
+  },
+});
+
+const sideloadedProgram2 = ZkProgram({
+  name: 'sideloadedProgram2',
+  publicInput: Field,
+  methods: {
+    recurseTwoSideloaded: {
+      privateInputs: [
+        SampleSideloadedProof,
+        VerificationKey,
+        SampleSideloadedProof2,
+        VerificationKey,
+      ],
+      async method(
+        publicInput: Field,
+        proof1: SampleSideloadedProof,
+        vk1: VerificationKey,
+        proof2: SampleSideloadedProof2,
+        vk2: VerificationKey
+      ) {
+        proof1.verify(vk1);
+        proof2.verify(vk2);
+
+        proof1.publicInput
+          .add(proof2.publicInput.field1.add(proof2.publicInput.field2))
+          .add(1)
+          .assertEquals(publicInput, 'PublicInput not matching');
+      },
+    },
+  },
+});
+
+describe('sideloaded', () => {
+  let program1Proof: Proof<Field, void>;
+  let program2Proof: Proof<Program2Struct, Field>;
+
+  let program1Vk: VerificationKey;
+  let program2Vk: VerificationKey;
+
+  before(async () => {
+    program1Vk = (await program1.compile()).verificationKey;
+    program2Vk = (await program2.compile()).verificationKey;
+
+    // Generate sample proofs
+    const proof1 = await program1.foo(Field(1), Field(1));
+    program1Proof = proof1;
+
+    const proof2 = await program2.foo(
+      { field1: Field(1), field2: Field(2) },
+      Field(3)
+    );
+    program2Proof = proof2;
+
+    await sideloadedProgram.compile();
+  });
```

Comment:
I find this `before()` pattern with variable reassignments unreadable and error-prone. I think it's the legacy of Jest + missing top-level await. Can you just move all that preparation logic to before the `describe()` block?